ext/libxml/libxml.c

```diff
@@ -1468,6 +1477,62 @@ PHP_LIBXML_API xmlChar *php_libxml_attr_value(const xmlAttr *attr, bool *free)
 	return value;
 }
 
+static zend_string *php_libxml_default_dump_doc_to_str(xmlDocPtr doc, int options, const char *encoding)
+{
+	xmlBufferPtr buf = xmlBufferCreate();
+	if (!buf) {
+		return NULL;
+	}
+
+	/* Encoding is handled from the encoding property set on the document */
+	xmlSaveCtxtPtr ctxt = xmlSaveToBuffer(buf, encoding, options);
+	if (!ctxt) {
+		xmlBufferFree(buf);
+		return NULL;
+	}
+
+	long status = xmlSaveDoc(ctxt, doc);
+	(void) xmlSaveClose(ctxt);
+	if (status < 0) {
+		xmlBufferFree(buf);
+		return NULL;
+	}
+
+	const xmlChar *content = xmlBufferContent(buf);
+	if (!content) {
+		xmlBufferFree(buf);
+		return NULL;
+	}
+
+	int size = xmlBufferLength(buf);
+	zend_string *str = zend_string_init((const char *) content, size, false);
+	xmlBufferFree(buf);
+	return str;
+}
+
+static zend_string *php_libxml_default_dump_node_to_str(xmlDocPtr doc, xmlNodePtr node, bool format, const char *encoding)
+{
+	// TODO: should this alloc take an encoding? For now keep it NULL for BC.
+	xmlOutputBufferPtr buf = xmlAllocOutputBuffer(NULL);
```

Comment:
I bet you'll love this.
Passing the encoding to the buffer would make it respect the actual encoding instead of just claiming it is that encoding.
I.e. now if you let SimpleXML or old DOM output a document, it will output the bytes in the encoding of the document, but if you output a node then it will output UTF-8 regardless of the document encoding.

Reply:
eeeeeee seems like using the document encoding makes the most sense, but this is a confusing API.